### DISCUSSION TITLE
for loop to fill a table

### DISCUSSION THREAD

**Discussion by AurelWeinhold** [typst/typst#3365] [Q&A] [answered]
Good evening,
I'm currently trying to fill a table with the elements in a dictionary by having the key be in the first column, and the value in the second column.
```typst
#table(
  columns: (1fr, 1fr),
  for pair in dict {
    [pair.at(0)]
    [pair.at(1)]
  }
)
``` 
This is how far I've gotten, but this does only provide content for the first column in every row. How do I split the for loop, so that it basically generates two content blocks, which are then interpreted by the table as the two different contents?

Thanks in advance
Greetings.

**Comment by PgBiel** [comment, marked as answer]
try:
```typ
#table(
  columns: (1fr, 1fr),
  ..for (k, v) in dict {
    (k, v)
  }
)
```

which is equivalent to

```typ
#table(
  columns: (1fr, 1fr),
  ..dict.pairs().map(((k, v)) => (k, v)).flatten()
)
```

though that's a no-op (you're mapping each pair to itself), so you can just

```typ
#table(
  columns: (1fr, 1fr),
  ..dict.pairs().flatten()  //  or ..for pair in dict { pair }
)
```

Note that this `..` syntax is the spread syntax, you can read more about it here: https://typst.app/docs/reference/foundations/arguments/#spreading

You use it to pass an entire array of values as arguments to a function. (Or an entire dictionary as named arguments.)

The for loop simply joins all of the returned values, so you get something like `(k, v) + (k, v) + (k, v) + ...`

In your sample code, it's joining content blocks, producing a single content block (certainly a single cell, since you didn't use spreading), through `[#k] + [#v] + [#k] + [#v] + ...` (each line is joined as well). What you need is to work with arrays to generate the cells in an array, and then spread the array, as demonstrated above.

**Reply by AurelWeinhold** [reply]
Thank you very much, that works like a charm!
